src/ray/gcs/gcs_server/node_info_handler_impl.cc

```diff
@@ -110,5 +110,89 @@ void DefaultNodeInfoHandler::HandleReportBatchHeartbeat(
                  << request.heartbeat_batch().batch_size();
 }
 
+void DefaultNodeInfoHandler::HandleGetResources(const GetResourcesRequest &request,
+                                                GetResourcesReply *reply,
+                                                SendReplyCallback send_reply_callback) {
+  ClientID node_id = ClientID::FromBinary(request.node_id());
+  RAY_LOG(DEBUG) << "Getting node resources, node id = " << node_id;
+
+  auto on_done = [node_id, reply, send_reply_callback](
+                     Status status,
+                     const boost::optional<gcs::NodeInfoAccessor::ResourceMap> &result) {
+    if (status.ok()) {
+      if (result) {
+        for (auto &resource : *result) {
+          (*reply->mutable_resources())[resource.first] = *resource.second;
+        }
+      }
+    } else {
+      RAY_LOG(ERROR) << "Failed to get node resources: " << status.ToString()
+                     << ", node id = " << node_id;
+    }
+    send_reply_callback(status, nullptr, nullptr);
+  };
+
+  Status status = gcs_client_.Nodes().AsyncGetResources(node_id, on_done);
+  if (!status.ok()) {
+    on_done(status, boost::none);
+  }
+
+  RAY_LOG(DEBUG) << "Finished getting node resources, node id = " << node_id;
+}
+
+void DefaultNodeInfoHandler::HandleUpdateResources(
+    const UpdateResourcesRequest &request, UpdateResourcesReply *reply,
+    SendReplyCallback send_reply_callback) {
+  ClientID node_id = ClientID::FromBinary(request.node_id());
+  gcs::NodeInfoAccessor::ResourceMap resources;
+  for (auto resource : request.resources()) {
+    resources[resource.first] = std::make_shared<rpc::ResourceTableData>(resource.second);
+  }
+
+  RAY_LOG(DEBUG) << "Updating node resources, node id = " << node_id;
+  auto on_done = [node_id, send_reply_callback](Status status) {
+    if (!status.ok()) {
+      RAY_LOG(ERROR) << "Failed to update node resources: " << status.ToString()
+                     << ", node id = " << node_id;
+    }
+    send_reply_callback(status, nullptr, nullptr);
+  };
+
+  Status status = gcs_client_.Nodes().AsyncUpdateResources(node_id, resources, on_done);
+  if (!status.ok()) {
+    on_done(status);
+  }
+
+  RAY_LOG(DEBUG) << "Finished updating node resources, node id = " << node_id;
+}
+
+void DefaultNodeInfoHandler::HandleDeleteResources(
+    const DeleteResourcesRequest &request, DeleteResourcesReply *reply,
+    SendReplyCallback send_reply_callback) {
+  ClientID node_id = ClientID::FromBinary(request.node_id());
+  std::vector<std::string> resource_names;
+  resource_names.reserve(request.resource_name_list_size());
+  for (int index = 0; index < request.resource_name_list_size(); ++index) {
+    resource_names.push_back(request.resource_name_list(index));
+  }
```

Comment:
there is a util function `MapFromProtobuf` for this.

Reply:
fixed